File: BTFViewer/web/src/utils/numericFormat.js
/**
 * Numeric presentation helpers (Step 3).
 * Lockstep with btf_viewer_pkg/numeric_format.py.
 *
 * Consistent percentile / duration formatting for Statistics tables and reports.
 */
import { formatTimeTrim } from './timeFormat.js'

/** Default decimal places for percentile columns. */
export const PERCENTILE_DECIMALS = 3

/**
 * Format p50/p95/p99/Max style duration values.
 * @param {number|null|undefined} value  Native trace units.
 * @param {string} scale  Trace timeScale.
 * @param {string} [kind] p50 | p95 | p99 | max | avg
 */
export function formatPercentile(value, scale, kind = '') {
  const v = Number(value)
  if (!Number.isFinite(v)) return '—'
  const k = String(kind || '').toLowerCase()
  if (k === 'max') return formatTimeTrim(v, scale)
  return formatTimeTrim(v, scale)
}

/** Right-align CSS class token for numeric table cells. */
export const NUMERIC_CELL_CLASS = 'num-cell'

/** Format a plain ratio as percentage with one decimal. */
export function formatRatioPct(ratio) {
  const v = Number(ratio)
  if (!Number.isFinite(v)) return '—'
  return `${(v * 100).toFixed(1)}%`
}

/** Format CPU utilisation 0–100 with consistent precision. */
export function formatCpuPct(pct) {
  const v = Number(pct)
  if (!Number.isFinite(v)) return '—'
  if (v >= 99.95) return '100%'
  if (v < 0.05 && v > 0) return '<0.1%'
  return `${v.toFixed(1)}%`
}

/**
 * Build an HTML table cell with numeric alignment.
 * @param {string} text
 * @param {string} [title]
 */
export function numericCellHtml(text, title = '') {
  const t = title ? ` title="${String(title).replace(/"/g, '&quot;')}"` : ''
  return `<td class="${NUMERIC_CELL_CLASS}"${t}>${text}</td>`
}

/**
 * Compare-friendly signed delta (Compare regression/improvement columns).
 * @param {number} delta  Native units; positive = regressed (worse).
 * @param {string} scale
 */
export function formatSignedDelta(delta, scale) {
  const v = Number(delta)
  if (!Number.isFinite(v) || v === 0) return '—'
  const sign = v > 0 ? '+' : '−'
  return `${sign}${formatTimeTrim(Math.abs(v), scale)}`
}

File: BTFViewer/btf_viewer_pkg/numeric_format.py
"""Numeric presentation helpers (Step 3).

Lockstep with ``web/src/utils/numericFormat.js``.
"""
from __future__ import annotations

from typing import Optional

PERCENTILE_DECIMALS = 3
NUMERIC_CELL_CLASS = "num-cell"


def _num_format_time_trim(value: float, scale: str) -> str:
    """Minimal parity with web formatTimeTrim.

    Named distinctly from ``timeline_util._format_time_trim`` on purpose: the
    monolithic bundle flattens every module into one namespace, so a second
    ``_format_time_trim`` here would shadow the real (unit-scaling) helper and
    every ``_build_trace_compare_rows`` cell would render as raw ``2.4e+06 us``.
    """
    try:
        return _format_time_trim(value, scale)  # timeline_util's, via bundle/bootstrap flatten
    except NameError:
        pass
    try:
        from .timeline_util import _format_time_trim as _impl  # noqa: WPS433

        return _impl(value, scale)
    except Exception:
        return f"{float(value):g} {scale}"


def format_percentile(value: Optional[float], scale: str, kind: str = "") -> str:
    if value is None:
        return "—"
    try:
        v = float(value)
    except (TypeError, ValueError):
        return "—"
    if not v and v != 0:
        return "—"
    return _num_format_time_trim(v, scale)


def format_ratio_pct(ratio: Optional[float]) -> str:
    if ratio is None:
        return "—"
    try:
        v = float(ratio)
    except (TypeError, ValueError):
        return "—"
    return f"{v * 100:.1f}%"


def format_cpu_pct(pct: Optional[float]) -> str:
    if pct is None:
        return "—"
    try:
        v = float(pct)
    except (TypeError, ValueError):
        return "—"
    if v >= 99.95:
        return "100%"
    if 0 < v < 0.05:
        return "<0.1%"
    return f"{v:.1f}%"


def numeric_cell_html(text: str, title: str = "") -> str:
    t = f' title="{title.replace(chr(34), "&quot;")}"' if title else ""
    return f'<td class="{NUMERIC_CELL_CLASS}"{t}>{text}</td>'


def format_signed_delta(delta: Optional[float], scale: str) -> str:
    if delta is None:
        return "—"
    try:
        v = float(delta)
    except (TypeError, ValueError):
        return "—"
    if v == 0:
        return "—"
    sign = "+" if v > 0 else "−"
    return f"{sign}{_num_format_time_trim(abs(v), scale)}"
